### PR TITLE
[coop] Remove unnecessary write barriers against thread locals and cleanup a little.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/metadata/handle.h [hunk 1]
```diff
@@ -401,19 +401,15 @@ This is why we evaluate index and value before any call to MONO_HANDLE_RAW or ot
 #define MONO_HANDLE_ARRAY_SETREF(HANDLE, IDX, VALUE) do {	\
 		int __idx = (IDX);	\
    		MonoObjectHandle __val = MONO_HANDLE_CAST (MonoObject, VALUE);		\
-		do {							\
-			MONO_HANDLE_SUPPRESS_SCOPE(1);			\
-			mono_array_setref_fast (MONO_HANDLE_RAW (MONO_HANDLE_UNSUPPRESS (HANDLE)), __idx, MONO_HANDLE_RAW (__val)); \
-		} while (0);						\
+		MONO_HANDLE_SUPPRESS_SCOPE(1);			\
```
Comment (lambdageek) [inline mono/metadata/handle.h]: Sorry, this needs to go back to the way it was - the way `MONO_HANDLE_SUPPRESS_SCOPE` is detected by Centrinel is a gross hack and it really has to be the first thing in a `{ compound statement }`

Comment (lambdageek) [reply mono/metadata/handle.h]: Alternately, I think wrapping the `mono_array_setref_fast` call in a `MONO_HANDLE_SUPPRESS()` expression would also work.

File: mono/metadata/handle.h [hunk 2]
```diff
@@ -401,19 +401,15 @@ This is why we evaluate index and value before any call to MONO_HANDLE_RAW or ot
...
 	} while (0)
 
 #define MONO_HANDLE_ARRAY_SETVAL(HANDLE, TYPE, IDX, VALUE) do {		\
 		int __idx = (IDX);					\
    		TYPE __val = (VALUE);					\
-		do {							\
-			MONO_HANDLE_SUPPRESS_SCOPE(1);			\
-			mono_array_set (MONO_HANDLE_RAW (MONO_HANDLE_UNSUPPRESS (HANDLE)), TYPE, __idx, __val); \
-		} while (0);						\
+		MONO_HANDLE_SUPPRESS_SCOPE(1);				\
```
Comment (lambdageek) [inline mono/metadata/handle.h]: same as setref - either `MONO_HANDLE_SUPPRESS_SCOPE(1)` as first thing inside braces, or else just wrap the `mono_array_set` in a a `MONO_HANDLE_SUPPRESS`

File: mono/metadata/handle.h (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/handle.h]: er, suppress vs. unsupress, nevermind that.

**Comment by lambdageek** [comment]
I made an issue for Centrinel to be more flexible.  Not sure when I'll get to it. https://github.com/lambdageek/centrinel/issues/21

File: mono/metadata/handle.h [hunk 3]
```diff
@@ -410,7 +411,8 @@ This is why we evaluate index and value before any call to MONO_HANDLE_RAW or ot
 #define MONO_HANDLE_ARRAY_SETVAL(HANDLE, TYPE, IDX, VALUE) do {		\
 		int __idx = (IDX);					\
    		TYPE __val = (VALUE);					\
-		do {							\
+		{	/* FIXME scope needed by Centrinel */		\
+			/* FIXME mono_array_set is not an expression. */ \
 			MONO_HANDLE_SUPPRESS_SCOPE(1);			\
 			mono_array_set (MONO_HANDLE_RAW (MONO_HANDLE_UNSUPPRESS (HANDLE)), TYPE, __idx, __val); \
 		} while (0);						\
```
Comment (lambdageek) [inline mono/metadata/handle.h]: you lost the "do {" that used to correspond to this "} while (0)"

File: mono/metadata/handle.h [hunk 4]
```diff
@@ -401,7 +401,8 @@ This is why we evaluate index and value before any call to MONO_HANDLE_RAW or ot
 #define MONO_HANDLE_ARRAY_SETREF(HANDLE, IDX, VALUE) do {	\
 		int __idx = (IDX);	\
    		MonoObjectHandle __val = MONO_HANDLE_CAST (MonoObject, VALUE);		\
-		do {							\
+		{	/* FIXME scope needed by Centrinel */		\
+			/* FIXME mono_array_setref_fast is not an expression. */ \
 			MONO_HANDLE_SUPPRESS_SCOPE(1);			\
 			mono_array_setref_fast (MONO_HANDLE_RAW (MONO_HANDLE_UNSUPPRESS (HANDLE)), __idx, MONO_HANDLE_RAW (__val)); \
 		} while (0);						\
```
Comment (lambdageek) [inline mono/metadata/handle.h]: This `} while (0)` is missing its `do {`

Comment (jaykrell) [reply mono/metadata/handle.h]: Oops right will fix tonight

File: mono/metadata/handle.h [hunk 5]
```diff
@@ -401,19 +401,21 @@ This is why we evaluate index and value before any call to MONO_HANDLE_RAW or ot
...
-		} while (0);						\
+		}							\
 	} while (0)
 
 #define MONO_HANDLE_ARRAY_SETVAL(HANDLE, TYPE, IDX, VALUE) do {		\
 		int __idx = (IDX);					\
    		TYPE __val = (VALUE);					\
-		do {							\
+		{	/* FIXME scope needed by Centrinel */		\
+			/* FIXME mono_array_set is not an expression. */ \
```
Comment (kumpera) [inline mono/metadata/handle.h]: I don't think we want mono_array_set(_fast) to be expressions.

Comment (jaykrell) [reply mono/metadata/handle.h]: Function calls are imho one the main things to model after and they are expressions. The macros look like function calls. I am I admit predisposed to want almost everything to be an expression w/o thinking about it much. You can build more on them. Even if they are void expressions. What should be a statement? Even things that are statements people complain about -- for example in some languages if/then is an expression -- kinda like a big ternary, returning their last expressions.

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Pull request merged**